### ISSUE TITLE
Messages parsed from a serialized form should know their serialized size

### ISSUE THREAD

**Issue by JanecekPetr** [protocolbuffers/protobuf#5729]
**What language does this apply to?**
Java (but the same idea can probably be introduced in all the other languages, too)

**Describe the problem you are trying to solve.**
We're trying to write protobufs to an output as fast as possible and we noticed they're computing their size even though the size should already be known because they were `parseDelimited()`'d from an `InputStream`.

Those protobufs have been read from a socket, filtered and now we're writing them. We expected that protobufs that have been `parseDelimited()` from an `InputStream` would know their size already as there is a size memoizing mechanism.

The problem is that apparently size computation takes around 50% of the writing time as suggested by this profile (heavily censored, sorry): [pump-60s.zip](https://github.com/protocolbuffers/protobuf/files/2865616/pump-60s.zip)

![screenshot here](https://i.imgur.com/A10GDnt.png)

**Describe the solution you'd like**
Protobufs read from a serialized form should have their size memoized so that if they're then written elsewhere, the size can be reused. If that's technically unfeasible for whatever reason, the messages read by `parseDelimited()` should know their size for sure.

**Describe alternatives you've considered**
Honestly, none. ~We can work around all this at the application level by remembering the size ourselves, but that is only that - a workaround~ (no we can't as there's no way to enforce the size on the message and it would be computed anyway).

**Additional context**
The profile above has been taken while using protobuf 2.6.1, but the same behaviour has been observed on 3.6.1, too.

**Comment by acozzette** [comment]
I think this is harder to do than it sounds. There is more than one way to represent a given message, so the size of the proto before it's parsed might be different from the output size after a parse and serialization cycle. For example, singular primitive fields can appear more than once and when you parse them the last one wins. So if you parse a message where a particular primitive field appears more than once, when you re-serialize it the size will get a little bit smaller since the duplicated field will just be serialized once. In your case if you're not modifying the message at all but just filtering out certain messages, is it possible to just hang onto the original serialized message and pass that along without re-serializing it?

**Comment by JanecekPetr** [comment]
> ...singular primitive fields can appear more than once and when you parse them the last one wins.

Oh, that's an interesting feature and yes, it defeats my idea.

> is it possible to just hang onto the original serialized message and pass that along without re-serializing it?

We'll have to explore this, yes. Unfortunately we'll need to copy the message out of the native memory onto the Java heap as we're reading from a `SocketChannel` via direct bytebuffers (*Btw - looking forward to 3.7.0's custom UTF-8 parsers for Java! Are any perf/alloc numbers available on it already, both on-heap and unsafe?*), and keep it around. We'll see if it's a worthwile thing to do. Probably not.

---

Here are a few more alternatives for completeness:

1. The naïve solution described above ("just remember the number of bytes read from the source") doesn't work because the number of bytes read from the source might not be indicative of the real size of the message once reserialized.

2. "It's an optimization, it does not have to work for those convoluted cases. Let's still use the idea, but bail out once we detect a singular field overwrite."

    That could work for proto2 where we have the information about field presence stored in a bitmasked field. The practicality of this would need to be determined (the run-time overhead could be nontrivial), though, and I'm already sceptical.

    For proto3 it simply does not work for primitive fields, at all. The optimization would have to therefore bail whenever a default value would be found on the stream, too. But those should be fairly rare and/or malicious:

    ```
    // ... in the message constructor
    int bytesReadBeforeParsing = input.getTotalBytesRead();

    // ... in the message constructor's parsing loop, an example of a singular string field:
    case 10: {
        java.lang.String s = input.readStringRequireUtf8();
        if ((memoizedSize == -2) || !stringField_.equals("") || s.isEmpty()) {
            // Bail. From now on we have no idea what the message size is.
            memoizedSize = -2;
        }
        stringField_ = s;
        break;
    }

    // ... and then, after the loop
    if (memoizedSize == -1) {
        // This message was well-behaved on the wire, we know its size:
        int bytesReadAfterParsing = input.getTotalBytesRead();
        memoizedSize = bytesReadAfterParsing - bytesReadBeforeParsing;
    } else {
        // restore the original "unknown" size
        memoizedSize = -1;
    }
    ```

    ...similar logic can be done for `Map` fields, too. I do not know enough about `oneOf` fields to know whether the approach is ok for them, too.

3. As an alternative, I conteplated per-field size bookkeeping while reading the message. Assuming default-valued fields are not written out, it could look like this:

    ```
    // ... in the message constructor's parsing loop, an example of a singular string field:
    case 10: {
        java.lang.String s = input.readStringRequireUtf8();
        if ((memoizedSize != -2) && (stringField_.equals("")) && !s.isEmpty()) {
            memoizedSize += input.getLastBytesRead();
            // getLastBytesRead() does not exist. If it's impractical to implement it,
            // we can impractically emulate it with two getTotalBytesRead() calls.
        } else {
            // Bail. From now on we have no idea what the message size is.
            memoizedSize = -2;
        }
        stringField_ = s;
        break;
    }
    ```

    Without much further thought, this could work. But I believe the previous solution is better, simpler.

    One downside of this (and the previous) approach is that as far as I can see it cannot be encapsulated into a custom/configurable `Parser` implementation, so it would need to be baked in, and opting-out is hard/impossible.

4. A completely new approach! Let's still use the original idea of reading the message as a whole, and storing into it the amount of bytes that were read during the message parsing. Later, during writing, if the write wrote less bytes in reality, it would pad the message with padding (zeros?) to make its size artifically bigger and conform to the "original" size.

    The upside: A new, independent API could be introduced, with things like `PaddedMessage` to encapsulate the new writing logic, and parsing can be "easily" done via a custom `Parser` implementation.

    The downside: It most likely terribly violates the wire format as no padding is allowed not expected by any of the current tooling afaik, so the data stream will appear to be corrupted/invalid to normal clients. Yay. :D

---

One additional problem to solve is that the `memoizedSize` field is currently not `volatile`, so it might not be visible later on. This needs to be properly taken care of.

Would you be interested in a deeper investigation into the option 2 - doing the optimization in the simple and clear cases only? The obvious thing to consider here is the runtime overhead for parsing. Is there any way to know whether anyone inside Google would benefit from the optimization?

**Comment by acozzette** [comment]
CC @ejona86 who knows more about Java performance

I don't really know about the UTF-8 parsers since I don't do much work on Java. About your option 2, I suspect that would be difficult to implement because there are multiple ways for the size to be different from what's expected. A singular field appearing multiple times is one example, but another one would be a non-canonical varint, e.g. a varint that would fit in one byte but happens to appear as two or more bytes on the wire. Varints are variable-length ints that can store small numbers in fewer bytes, but nothing stops you from making them take up more bytes than necessary.

There's a C++ library whose name I've now forgotten which actually takes advantage of non-canonical varints to speed up serialization. It uses 4 bytes for every size-prefix varint. That imposes a size maximum of 256 MiB, but that's plenty for typical use cases. The idea is that it only has to make one pass over the message, because it can set aside 4 bytes for each size prefix and just fill those in later on once the size has been calculated. This way the serialization can happen in one pass without requiring an earlier pass for computing sizes. This is similar to your option 4 but is a valid way of including extra padding.

I know there was also an attempt to implementation serialization in reverse order from back to front, so that by the time we're ready to write the size prefix we will have already calculated it. I'm not sure if this idea went anywhere, though.

These ideas are all kind of interesting to think about, though you may want to be careful about spending too much time on it because this sort of thing is a fairly deep change that will likely be difficult to do. I think it's the kind of thing where if it were easy then someone would have already done it. If serialization is really a bottleneck for your use case, you might want to think about figuring out ways to just do less serialization, or even switch to a  different wire format like FlatBuffers or Cap'n Proto. From what I understand those formats are a little bit harder to use but have the advantage of having almost "free" parsing and serialization, since the wire format and in-memory format are exactly the same.

**Comment by ejona86** [comment]
> I know there was also an attempt to implementation serialization in reverse order from back to front, so that by the time we're ready to write the size prefix we will have already calculated it. I'm not sure if this idea went anywhere, though.

There were some benchmarks and they did show some improvement, although it was hard to make varint itself as fast. But it is blocked on the table-driven parser.

**Comment by JanecekPetr** [comment]
> non-canonical varint

Nasty.

> I think it's the kind of thing where if it were easy then someone would have already done it.

That's exactly the sentiment I had when creating the ticket, and yet I did obviously not see the potential complications.

Thank you for your time and your clarifications. I believe the option 3 outlined above is the only possible implementation of my idea, and yet it is totally not worth it and would mean a solid perf regression for "normal" users.

I think we can close this issue with a clean consciense.

---

For anyone solving the same problem in the future:

We'll try to solve this by implementing our own delegating `Parser` that will reflectively call `setMemoizedSerializedSize()` for trusted sources. The performance of this solution is to be determined, as well as the multi-threaded friendliness.

And no, we do not think that the protobuf project should expose any of the machinery required to do this in a cleaner way.
